pennylane/labs/trotter_error/abstract.py

```diff
@@ -0,0 +1,167 @@
+    def __mul__(self, scalar: float) -> Fragment:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __matmul__(self, other: Fragment) -> Fragment:
+        raise NotImplementedError
+
+    @abstractmethod
+    def norm(self, params: Dict) -> float:
+        """Compute the norm of the fragment.
+
+        Args:
+            params (Dict): A dictionary of parameters needed to compute the norm. It should be
+                specified for each class inheriting from :class:`~.Fragment`.
+
+        Returns:
+            float: the norm of the :class:`~.Fragment`
+
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def apply(self, state: AbstractState) -> AbstractState:
+        """Apply the Fragment to a state on the right. The type of ``state`` is determined by each class inheriting from ``Fragment``.
+
+        Args:
+            state (AbstractState): an object representing a quantum state
+
+        Returns:
+            AbstractState: the result of applying the ``Fragment`` to ``state``
+
+        """
+        raise NotImplementedError
+
+    def expectation(self, left: AbstractState, right: AbstractState) -> float:
+        """Return the expectation value of a state. The type of ``state`` is determined by each class inheriting from ``Fragment``.
+
+        Args:
+            left (AbstractState): the state to be multiplied on the left of the ``Fragment``
+            right (AbstractState): the state to be multiplied on the right of the ``Fragment``
+
+        Returns:
+            float: the expectation value obtained by applying ``Fragment`` to the given states
+        """
+        return left.dot(self.apply(right))
+
+
+def commutator(a: Fragment, b: Fragment) -> Fragment:
+    """Return the commutator of two :class:`~.Fragment` objects
+
+    Args:
+        a (Fragment): the :class:`~.Fragment` on the left side of the commutator
+        b (Fragment): the :class:`~.Fragment` on the right side of the commutator
+
+    Returns:
+        Fragment: the commutator ``[a, b]``
+    """
+    return a @ b - b @ a
+
+
+def nested_commutator(fragments: Sequence[Fragment]) -> Fragment:
+    """Return the nested commutator of a sequence of :class:`~.Fragment` objects
+
+    Args:
+        fragments (Sequence[Fragment]): a sequence of fragments
+
+    Returns:
+        Fragment: the nested commutator of the fragments
+    """
+
+    if len(fragments) < 2:
+        raise ValueError("Need at least two fragments to commute.")
+
+    if len(fragments) == 2:
+        return commutator(*fragments)
+
+    head, *tail = fragments
+
+    return commutator(head, nested_commutator(tail))
+
+
+class AbstractState(ABC):
+    """Abstract class defining the methods a class needs to implement to be used to compute an expectation value.
+
+    A class inheriting from ``AbstractState`` must implement the following dunder methods.
+
+    * ``__add__``: implements addition
+    * ``__sub__``: implements subtraction
```

Comment:
I think we can skip this as we get `__sub__` from `__add__` and `__mul__` themselves?